keras/src/backend/jax/trainer.py

```diff
@@ -884,75 +886,44 @@ def jax_state_sync(self):
                 ref_v.assign(v)
         self._jax_state_synced = True
 
-    def _record_training_state_sharding_spec(self):
-        self._trainable_variable_shardings = [
-            v.value.sharding for v in self.trainable_variables
-        ]
-        self._non_trainable_variable_shardings = [
-            v.value.sharding for v in self.non_trainable_variables
-        ]
-        if hasattr(self, "optimizer") and self.optimizer is not None:
-            self._optimizer_variable_shardings = [
-                v.value.sharding for v in self.optimizer.variables
-            ]
-        else:
-            self._optimizer_variable_shardings = []
-        self._metrics_variable_shardings = [
-            v.value.sharding for v in self.metrics_variables
-        ]
-
-    def _clear_jax_state_sharding(self):
-        self._trainable_variable_shardings = None
-        self._non_trainable_variable_shardings = None
-        self._optimizer_variable_shardings = None
-        self._metrics_variable_shardings = None
-
-    def _enforce_jax_state_sharding(
-        self,
-        trainable_variables=None,
-        non_trainable_variables=None,
-        optimizer_variables=None,
-        metrics_variables=None,
-    ):
-        """Enforce the sharding spec constraint for all the training state.
-
-        Since the output of the train/eval step will be used as inputs to next
-        step, we need to ensure that they have the same sharding spec, so that
-        nnx.jit/jax.jit won't have to recompile the train/eval function.
-
-        Note that this function will also rely on the recorded sharding spec
-        for each of states.
+    def _get_training_state_shardings(self):
+        distribution = distribution_lib.distribution()
+        mesh = distribution.device_mesh.backend_mesh
+
+        def get_sharding(variable):
```

Comment:
I don't understand why you need `get_sharding`.

Previously, it was simply doing `v.value.sharding`, does that no longer work?

Additionally, this is how `KerasVariable` does it: https://github.com/keras-team/keras/blob/master/keras/src/backend/jax/core.py#L38-L45
One important part is to *not* use `distribution` if `variable._layout` is not `None`, we have an important use case in KerasRS where the distribution of some variables is done separately.